src/libbson/src/bson/bson-utf8.c

```diff
@@ -230,6 +230,118 @@ bson_utf8_validate (const char *utf8, /* IN */
    return true;
 }
 
+/*
+ *--------------------------------------------------------------------------
+ *
+ * _is_special_char --
+ *
+ *       Uses a bit mask to check if a character requires special formatting
+ *       or not. Called from bson_utf8_escape_for_json.
+ *
+ * Parameters:
+ *       @c: An unsigned char c.
+ *
+ * Returns:
+ *       true if @c requires special formatting. otherwise false.
+ *
+ * Side effects:
+ *       None.
+ *
+ *--------------------------------------------------------------------------
+ */
+
+static BSON_INLINE bool
+_is_special_char (unsigned char c)
+{
+   /*
+   C++ equivalent:
+   std::bitset<256> charmap = [...]
+   return charmap[c];
+   */
+   static const bson_unichar_t charmap[8] = {0xffffffff, // control characters
+                                             0x00000004, // double quote "
+                                             0x10000000, // backslash
+                                             0x00000000,
+                                             0xffffffff,
+                                             0xffffffff,
+                                             0xffffffff,
+                                             0xffffffff}; // non-ASCII
+   const int int_index = c / (sizeof (bson_unichar_t) * 8);
+   const int bit_index = c & (sizeof (bson_unichar_t) * 8 - 1);
+   return (charmap[int_index] >> bit_index) & 1;
+}
+
+/*
+ *--------------------------------------------------------------------------
+ *
+ * _bson_utf8_handle_special_char --
+ *
+ *       Appends a special character in the correct format when converting
+ *       from UTF-8 to JSON. This includes characters that should be escaped
+ *       as well as ASCII control characters.
+ *
+ *       Normal ASCII characters are handled in bson_utf8_escape_for_json,
+ *       where this function is called from.
+ *
+ * Parameters:
+ *       @c: A bson_unichar_t.
+ *       @str: A string to append the special character to.
+ *
+ * Returns:
+ *       None.
+ *
+ * Side effects:
+ *       None.
+ *
+ *--------------------------------------------------------------------------
+ */
+
+static BSON_INLINE void
+_bson_utf8_handle_special_char (bson_unichar_t c,   /* IN */
+                                bson_string_t *str) /* OUT */
+{
+   BSON_ASSERT (c < 0x80);
+   BSON_ASSERT (str);
+
+   switch (c) {
+   case '"':
+      bson_string_append (str, "\\\"");
+      break;
+   case '\\':
+      bson_string_append (str, "\\\\");
+      break;
+   case '\b':
+      bson_string_append (str, "\\b");
+      break;
+   case '\f':
+      bson_string_append (str, "\\f");
+      break;
+   case '\n':
+      bson_string_append (str, "\\n");
+      break;
+   case '\r':
+      bson_string_append (str, "\\r");
+      break;
+   case '\t':
+      bson_string_append (str, "\\t");
+      break;
+   default: {
+      // ASCII control character
+      BSON_ASSERT (c < 0x20);
```

Comment:
```suggestion
      BSON_ASSERT (c < 0x20u);
```
Signedness consistency.